services/galley/src/Galley/API/Teams.hs

```diff
@@ -856,141 +861,207 @@ canUserJoinTeam tid = do
       when (size >= limit) $ do
         throwM tooManyTeamMembersOnTeamWithLegalhold
 
--- Public endpoints for feature checks
-
-getFeatureStatusH :: UserId ::: TeamId ::: Public.TeamFeatureName ::: JSON -> Galley Response
-getFeatureStatusH (uid ::: tid ::: featureName ::: _) =
-  json <$> getFeatureStatus uid tid featureName
-
-getFeatureStatus :: UserId -> TeamId -> Public.TeamFeatureName -> Galley Public.TeamFeatureStatus
-getFeatureStatus uid tid featureName = do
-  zusrMembership <- Data.teamMember tid uid
-  void $ permissionCheck (ViewTeamFeature featureName) zusrMembership
-  getFeatureStatusInternal tid featureName
-
--- | Get feature flag status for a team.  To be called only from authorized personnel (e.g.,
--- from a backoffice tool)
-getFeatureStatusInternalH :: TeamId ::: Public.TeamFeatureName ::: JSON -> Galley Response
-getFeatureStatusInternalH (tid ::: featureName ::: _) = do
-  json <$> getFeatureStatusInternal tid featureName
-
-getFeatureStatusInternal :: TeamId -> Public.TeamFeatureName -> Galley Public.TeamFeatureStatus
-getFeatureStatusInternal tid featureName = do
-  case featureName of
-    Public.TeamFeatureLegalHold -> getLegalholdStatusInternal tid
-    Public.TeamFeatureSSO -> getSSOStatusInternal tid
-    Public.TeamFeatureSearchVisibility -> getTeamSearchVisibilityAvailableInternal tid
-    Public.TeamFeatureValidateSAMLEmails -> getValidateSAMLEmailsInternal tid
-    Public.TeamFeatureDigitalSignatures -> getDigitalSignaturesInternal tid
-
--- | Enable or disable feature flag for a team.  To be called only from authorized personnel
--- (e.g., from a backoffice tool)
-setFeatureStatusInternalH :: TeamId ::: Public.TeamFeatureName ::: JsonRequest Public.TeamFeatureStatus ::: JSON -> Galley Response
-setFeatureStatusInternalH (tid ::: featureName ::: req ::: _) =
-  (empty & setStatus status204) <$ (setFeatureStatusInternal tid featureName =<< fromJsonBody req)
-
-setFeatureStatusInternal :: TeamId -> Public.TeamFeatureName -> Public.TeamFeatureStatus -> Galley ()
-setFeatureStatusInternal tid featureName status = do
-  case featureName of
-    Public.TeamFeatureLegalHold -> setLegalholdStatusInternal tid status
-    Public.TeamFeatureSSO -> setSSOStatusInternal tid status
-    Public.TeamFeatureSearchVisibility -> setTeamSearchVisibilityAvailableInternal tid status
-    Public.TeamFeatureValidateSAMLEmails -> setValidateSAMLEmailsInternal tid status
-    Public.TeamFeatureDigitalSignatures -> setDigitalSignaturesInternal tid status
-
-getSSOStatusInternal :: TeamId -> Galley Public.TeamFeatureStatus
+data FeatureStatusHandlers (a :: Public.TeamFeatureName) = FeatureStatusHandlers
+  { fshGet :: UserId ::: TeamId ::: JSON -> Galley Response,
+    fshGetInternal :: TeamId ::: JSON -> Galley Response,
+    fshSet :: UserId ::: TeamId ::: JsonRequest (Public.TeamFeatureStatus a) ::: JSON -> Galley Response,
+    fshSetInternal :: TeamId ::: JsonRequest (Public.TeamFeatureStatus a) ::: JSON -> Galley Response
+  }
```

Comment:
What about this?

```
setFeatureStatus :: DoAuth {- custom bool -] -> Maybe UserId -> TeamId -> Gally Response
getFeatureStatus :: DoAuth {- custom bool -] -> Maybe UserId -> TeamId -> Gally Response
```

Then we could switch between internal and external semantics on the call side.

Reply:
Done